compute the minimum distance of QLDPC using Mixed Integer Programming

Fe-r-oz opened the pull request:
This PR implements the Mixed Integer programming (MIP) approach to compute the minimum distance of QLDPC using GNU's Linear Programming Kit. This method is used by [Panteleev and Kalachev]( https://arxiv.org/pdf/1904.02703) and by [Bravyi et. al.](https://arxiv.org/pdf/2308.07915). The latter reference mentioned that this MIP approach was originally developed in 2011 and used in this [reference](https://arxiv.org/pdf/1108.5738)

This PR allows the users to compute the minimum distance with ease and also provide the necessary documentation along with the relevant modern usecases.  The tests have been conducted in #435

ILP/MILP methods: https://quantumcomputing.stackexchange.com/questions/37289/compute-the-exact-minimum-distance-of-a-qecc-with-integer-linear-programming-met

P.S. Mixed Integer Programming (MIP) or mixed integer linear program refer to same thing in the literature, in this context.

Edit:
- [x]  The code is properly formatted and commented.
- [x] Substantial new functionality is documented within the docs.
- [x] All new functionality is tested.
- [x] All of the automated tests on github pass.
- [x] We recently started enforcing formatting checks. If formatting issues are reported in the new code you have written, please correct them.

Fe-r-oz commented:
Please help review this PR, thank you :)

Thanks to GNU for open-souring [GLPK](https://www.gnu.org/software/glpk/). I have tested all the instances of 2bga codes using `distance` and added other tests for correctness. Many more tests are in #435. Referring to #353 as it's related a issue. This PR aims to provide the first reasonable step. This MIP method is used by [Panteleev and Kalachev](https://arxiv.org/pdf/1904.02703) and by [Bravyi et. al.](https://arxiv.org/pdf/2308.07915). This MILP method that Bravyi used was also asked on [stack exchange]( https://quantumcomputing.stackexchange.com/questions/37289/compute-the-exact-minimum-distance-of-a-qecc-with-integer-linear-programming-met
). The end goal is to be able to compute minimum distance of QECC/QLDPC codes with n being $10^3$ or $10^4$: QDistRnd or MaxSAT solvers or upcoming algorithms appear to be helpful for that. There is a lot more work to be done in this direction. I will be happy if this PR is one of my PRs you are most excited about, along with in-place Ted's Pauli measurement algorithm PR. :)

Possible Future TODOs include:

- [ ] More performance optimizations if possible
- [ ] comparison with distance solvers from stim. 
  - [ ] It uses Microsoft's [z3](https://github.com/Z3Prover/z3) SAT solver. This method formulate task the problem into maxSAT problem. Julia does have the [z3.jl](https://github.com/ahumenberger/Z3.jl)binding! stim also uses pysat for maxSAT problem.
   - [ ] stim also uses [pysat](https://github.com/pysathq/pysat) to convert distance-finding problem into a maxSAT problem
- [ ] comparison with QDistRnd Package from GAP: we can use Oscar to directly use QDistRnd package. User would have to provide the matrix is MTX format.
- [ ] integrate latest solvers: Grassl, Eric and others are working towards it and have some papers in this new directions
- [ ] implement minimum distance solvers in pure Julia
- [ ] Benchmark Lifted product codes #356
- [ ] Benchmark 2BGA codes with finitely presented groups and with Direct product of General Groups, e.g. direct product of two finite groups using Oscar #400
- [ ] Benchmark #388,
- [ ] Benchmark #378
- [ ] Benchmark #381
- [ ] Benchmark #399
- [ ] Benchmark #392
- [ ] Benchmark #436
- [ ] Benchmark #441
- [ ] Benchmark #442
- [ ] Benchmark color codes with varying distances according to formula

Krastanov commented:
This is really interesting and deserving of a bounty.

Regrettably there are a few merge conflicts related to the doc fixes in an earlier PR. Could you fix those?

I also probably would suggest looking into a slightly different API, or at least some easier way to support independent solvers. But that can be discussed after the first round of review.

Krastanov commented:
I will mark it as a draft, just to make managing my review queue a bit easier. Just re-request review and mark it as ready after the fixes.

Fe-r-oz commented:
> Some easier way to support independent solvers.

That's a great suggestion `(solver = )`. I have been reading the JuMP discourse recently about MIP. User can use `HiGHs.jl`, `GLPK.jl`, the underlying solver for these is free, but `Gurobi.jl` or `CPLEX.jl` have some underlying solver that require obtaining license. If people want to use closed sourced solvers for faster performance maybe, we can give them that choice. In addition, I found these posts interesting:

- https://discourse.julialang.org/t/is-glpk-still-maintained-relevant/116950/4
- https://discourse.julialang.org/t/fastest-open-licensed-solver-for-integer-problems/17440
- https://discourse.julialang.org/t/error-with-highs-independent-parallel-solves/96048/2
- https://stackoverflow.com/questions/502102/best-open-source-mixed-integer-optimization-solver

From `Gurobi.jl` documentation: `The underlying solver is a closed-source commercial product for which you must [obtain a license](https://www.gurobi.com/). Free Gurobi licenses are available for [academics and students](https://www.gurobi.com/academia/academic-program-and-licenses/).`

As a future TODO: Certain MLE decoders require MIP solvers as well.  This week, [Correlated decoding of logical algorithms with transversal gate](https://arxiv.org/pdf/2403.03272) appeared on arXiv, and it used MLE decoder that use MIP approach. This paper references the same 2011 paper that introduced these MLE decoders using mixed-integer-program.

Fe-r-oz commented:
Thank you very much for your valuable feedback. The code is now independent of any specific MIP solver.  

I chose `HiGHS` as the default MIP solver over `GLPK` because of the ongoing collaboration between `JuMP` and `HiGHS` to enhance the `HiGHS.jl` package. This makes `HiGHS` a more forward-looking choice.  

I have not added `SCIP.jl` as a solver yet due to dependency issues. Specifically, `SCIP.jl` relies on `polymake`, which conflicts with other dependencies like `Hecke` or `Nemo`. Addressing this is on my future TODO list.  

Regarding multi-threading, `GLPK` does not support it, and `HiGHS` currently has some issues with independent parallel solves: [JuliaLang Discourse discussion](https://discourse.julialang.org/t/error-with-highs-independent-parallel-solves/96048/2). Once these issues are resolved, we can revisit adding multi-threading support.  

I look forward to hearing your thoughts and suggestions regarding the API.  Thank you once again!

Fe-r-oz commented:
Please help review this PR, thank you!

Fe-r-oz commented:
>I also probably would suggest looking into a slightly different API, or at least some easier way to support independent solvers

Thank you once again for this helpful insight.  After some more introspection, I am turning this back to draft status.  This is because today, I came across the latest [JuMP discourse post](https://discourse.julialang.org/t/ci-integration-failed-with-gurobi-jl-in-environment/126236/2?u=fe-r-oz) and insightful comment by Oscar: _Let the user choose which solver to use. There is no need to add it to the Project.toml. Adding solvers to the Project.toml means that every user installs all of the solvers, even if they never end up using them, and it runs into license issues like the one you are experiencing._

This reminded me of your insightful comment.  I think this is what you meant by "easier way to support independent solvers"! It seems that I didn't understood it correctly. Currently, I have  I added `QuantumCliffordJuMPExt = ["GLPK", "HiGHS", "JuMP"]` in `project.toml` but everything works fine and okay 🙂 I have used `opt = solver.Optimizer` with default `solver = HIGHS`, and then I call  `model = Model(opt; add_bridges = false).` 

Based on the insights gathered, the user can decide which solver to pass as `solver.Optimizer`? That way, there would be no need to have `GLPK, HiGHS` or any other solver in the `project.toml`.

My apologies for the mistake! I'm glad that your insightful comment reminded me of this!

Fe-r-oz commented:
> I also probably would suggest looking into a slightly different API, or at least some easier way to support independent solvers

The implementation is now independent of solvers, both closed source or open source MIP solvers can now be used! 

This PR is now ready for review, Thank you!

Krastanov commented on ext/QuantumCliffordJuMPExt/min_distance_mixed_integer_programming.jl:
```diff
@@ -0,0 +1,226 @@
+    @constraint(model, sum(weight[i] * x[i] for i in 1:num_var) == 1)
+    optimize!(model)
+    # Ensure the model is solved and feasible
+    is_solved_and_feasible(model) || error("Did not solve model")
```
use a throw with a specific error type

Krastanov commented on test/test_ecc_2bga.jl:
```diff
@@ -1,7 +1,9 @@
 @testitem "ECC 2BGA" begin
```
double check that the extra time these tests take is no more than a minute

Fe-r-oz replied:
It takes 52.2 seconds!

```
Test Summary:                    | Pass  Total   Time
Reproduce Table 2 lin2024quantum |   32     32  52.2s
Test.DefaultTestSet("Reproduce Table 2 lin2024quantum", Any[], 32, false, false, true, 1.748086602231193e9, 1.748086654395902e9, false, "REPL[11]")

```

Krastanov commented on ext/QuantumCliffordJuMPExt/min_distance_mixed_integer_programming.jl:
```diff
@@ -0,0 +1,226 @@
+the code distance was calculated using the mixed integer programming approach.
+
+"""
+function distance(c::Stabilizer; upper_bound=false, logical_qubit=code_k(c), all_logical_qubits=false, logical_operator_type=:X, solver::Module)
```
```suggestion
function distance_mip(c::Stabilizer; upper_bound=false, logical_qubit=code_k(c), all_logical_qubits=false, logical_operator_type=:X, solver::Module)
```

If we have this different name, it is easier to distinguish the two approaches (database vs MIP algorithm).

The error message for `distance` can get a hint on `MethodError` to direct people to try this function instead.

Krastanov replied:
another options is to do `distance(code, alg::AbstractDistanceAlg)` and for instance it is then `distance(code, DistanceDataBase())` and `distance(code, DistanceMIPAlgorithm(configuraiton_options))`.

I think I like this option better.

And then we can have a default:
```
function distance(code)
    try
        distance(code, DistanceDatabase())
    catch MethodError
        if it is method error about distance
        throw(NotInDatabase("it is not in the database......... consider using `distance(code, WhateverIsTheNameOfTheOtherMethod)`"))
...
```

Fe-r-oz replied:
Thank you!! suggestion incorporated hopefully!!

Fe-r-oz commented on ext/QuantumCliffordJuMPExt/min_distance_mixed_integer_programming.jl:
```diff
@@ -0,0 +1,226 @@
+    logical_operator_type == :X || logical_operator_type == :Z || throw(ArgumentError("Invalid type of logical operator: Use :X or :Z"))
+    l = get_lx_lz(c)[1]
+    H = SparseMatrixCSC{Int, Int}(stab_to_gf2(parity_checks(c)))
+    h = get_stab(H, :X)
```
Thank you very much!! ✨ 🙂  The approach suggested is absolutely correct! Indeed, due to CSS construction, the missing columns are zeros so we can just use 'cat'.`

```julia
julia> using QuantumClifford.ECC

julia> c = parity_checks(Steane7());

julia> stab_to_gf2(c)
6×14 Matrix{Bool}:
 0  0  0  1  1  1  1  0  0  0  0  0  0  0
 0  1  1  0  0  1  1  0  0  0  0  0  0  0
 1  0  1  0  1  0  1  0  0  0  0  0  0  0
 0  0  0  0  0  0  0  0  0  0  1  1  1  1
 0  0  0  0  0  0  0  0  1  1  0  0  1  1
 0  0  0  0  0  0  0  1  0  1  0  1  0  1

julia> lx = stab_to_gf2(QuantumClifford.ECC.logx_ops(c))
1×14 Matrix{Bool}:
 0  0  1  0  1  1  0  0  0  0  0  0  0  0

julia> lz = stab_to_gf2(QuantumClifford.ECC.logz_ops(c))
1×14 Matrix{Bool}:
 0  0  0  0  0  0  0  0  1  0  1  0  1  0

julia> parity_checks_x(Steane7())
3×7 Matrix{Bool}:
 0  0  0  1  1  1  1
 0  1  1  0  0  1  1
 1  0  1  0  1  0  1

julia> parity_checks_z(Steane7())
3×7 Matrix{Bool}:
 0  0  0  1  1  1  1
 0  1  1  0  0  1  1
 1  0  1  0  1  0  1

julia> x_part = [px zeros(Bool, size(px, 1), 7)]
3×14 Matrix{Bool}:
 0  0  0  1  1  1  1  0  0  0  0  0  0  0
 0  1  1  0  0  1  1  0  0  0  0  0  0  0
 1  0  1  0  1  0  1  0  0  0  0  0  0  0

julia> x_part = cat(px, spzeros(Bool, size(px, 1), 7); dims=2)
3×14 SparseMatrixCSC{Bool, Int64} with 12 stored entries:
 ⋅  ⋅  ⋅  1  1  1  1  ⋅  ⋅  ⋅  ⋅  ⋅  ⋅  ⋅
 ⋅  1  1  ⋅  ⋅  1  1  ⋅  ⋅  ⋅  ⋅  ⋅  ⋅  ⋅
 1  ⋅  1  ⋅  1  ⋅  1  ⋅  ⋅  ⋅  ⋅  ⋅  ⋅  ⋅
```

Therefore, we use the following and remove the weird `get_stab(H, :X)` function.

```julia
px = SparseMatrixCSC{Int, Int}(parity_checks_x(c))
h = cat(px, spzeros(Int, size(px, 1), nqubits(c)); dims=2)
```

Fe-r-oz commented on ext/QuantumCliffordJuMPExt/min_distance_mixed_integer_programming.jl:
```diff
@@ -0,0 +1,226 @@
+    wlx = count(!iszero, lx) # Count non-zero elements in lx
+    # Number of slack variables needed to express orthogonality constraints modulo two
+    num_anc_hx = ceil(Int, log2(whx))
+    num_anc_logical = ceil(Int, log2(wlx))
```
Thank you!! done!!

Fe-r-oz commented:
Thank you very much for the amazing improvements and suggestions on the documentation build!! Nothing appears to be broken by the changes.

Regarding LaTeX rendering: I tried building the documentation locally, but since I don’t yet have VSCode installed, I didn’t get the HTML link of documentation after processing the commands. For LaTeX, I used `@doc raw"""` to avoid escaping `\` or `$`, which hopefully seems to work well.

A few other fixes to the aforementioned issues: resolved the formatting error with `!!! note`, added the missing citation and followed the `QuantumSavory.ProtocolZoo` format for the `MIPDistanceAlgorithm` docstring.

The CI documentation error is unrelated to this PR—it occurs because `linkcheck` sometimes fails when testing tutorial Jupyter notebook link. Please help review this PR, Thank you!!

Krastanov commented:
> Regarding LaTeX rendering: I tried building the documentation locally, but since I don’t yet have VSCode installed, I didn’t get the HTML link of documentation after processing the commands

Not immediately important, but just posting for future reference: VS Code should not be related or necessary for the doc building. You probably did not get the `index.html` and others because your doc build failed. If that is the case, it will report why it failed. You can just add exceptions to those failures in the config in `main.jl`, e.g. `warnonly = [:some_other_reason, ...]` and `checklinks=false`.

Krastanov reviewed (changes requested):
Thanks! This looks great. One last comment left below.

After you address that comment, please go ahead and officially claim the bounty, no need to wait for review. It would be convenient for me if you make the claim promptly, as otherwise some of the processing can be significantly delayed.

Fe-r-oz commented on src/ecc/ECC.jl:
```diff
@@ -123,23 +123,35 @@ end
+    logical operators of the specified type (`X` or `Z`)"""
     all_logical_qubits::Bool=false
+    """type of logical operator to consider (:X or :Z, defaults to :X) - both types yield identical distance results for CSS stabilizer codes."""
     logical_operator_type::Symbol=:X
```
Absolutely agree!! Thank you very much!!

Fe-r-oz commented:
> VS Code should not be related or necessary for the doc building.

Thank you for your amazing suggestions!! That's awesome news!!

##

Thank you very much for helpful suggestion!! Incorporated the code review suggestion and added remaining distance test for the remaining GB codes in (`test_ecc_generalized_bicycle_codes.jl`) so that the entire testset there does not take more than extra 1 minute. When I test all the GB codes there, it takes total 2 minutes to run the entire set of codes so I left one of them out so extra time < 1 min. Testing individually, it takes ≤ 60s to test distance of each of these GB codes though, within default time limit of solver. Please let me know if there’s anything else that needs to be done!! Thank you!!

Fe-r-oz commented:
Hi, @Krastanov, just a friendly bump on this PR! Thank you!

Fe-r-oz commented:
Thank you! 😃
